Improve configuration documentation (#498)
transparency-dev/tesseract: docs/configuration.md

TesseraCT responses to `add-*` requests as soon as Tessera has assigned a
sequence number to the corresponding entry. The entry would then get integrated
and published afterwards, which might happen before TesseraCT responses to the
`add-*` request, or after.

This flag is on by default.

### Sequencing Batch

The `batch_max_age` and `batch_max_size` flags control the maximum age and number
of entries in a single [sequencing batch](https://github.com/transparency-dev/tessera?tab=readme-ov-file#sequencing).
These flags affect the overall performance of TesseraCT. The correct values for
these flags depend on multiple factors, such as the driver used, the number
of TesseraCT servers or their steady QPS rate.

## Antispam

The `pushback_max_antispam_lag`, `pushback_max_dedupe_in_flight` and
`inmemory_antispam_cache_size` flags control how [TesseraCT's Antispam
feature](./architecture.md#antispam) works, which itself is built on top of
[Tessera's Antispam](https://github.com/transparency-dev/tessera?tab=readme-ov-file#antispam)
capabilities. It is composed of three main steps:

1. A process **asynchronous** with `add-*`, populating the antispam databases

> with entry indexes as the log grows. This is handled by Tessera.

Reviewer: ```suggestion
with entry indices as the log grows. This is handled by Tessera.
```

2. A call **synchronous** with **all** `add-*` to the antispam database to
identify if the entry submitted is a duplicate of a previous one. This is
handled by Tessera.
3. If this entry is a duplicate, a last call **synchronous** with this duplicate
`add-*` call fetches the entry from the log to extract the information required
to build its SCT. This is handled by TesseraCT.

> The `pushback_max_antispam_lag` flag controls the maximum number of entries that
> have already been integrated into the log, but that have not been added to the
> antispam database yet by the `1.` **asynchronous** process. When this value is
> exceeded, TesseraCT returns `429 - Too Many Requests` to **all** `add-*`
> requests. This protects TesseraCT from accepting new entries without being able
> to tell whether they are duplicates of recently integrated entries.

Reviewer: ```suggestion
The `pushback_max_antispam_lag` flag controls the limit of how far behind the current size of the tree the asynchronous process in (1) can fall.
When this value is exceeded, TesseraCT returns `429 - Too Many Requests` to **all** `add-*`requests. This protects TesseraCT from accepting new entries without being able to determine whether they are duplicates of recently integrated entries.
```

The `inmemory_antispam_cache_size` flag controls the number of entries from this
database are cached locally on each TesseraCT server, to make subsequent `2.`
calls faster.

The `pushback_max_dedupe_in_flight` flag rate limits how many concurrent `add-*`
requests identified as duplicates can be processed by the
`3.` **synchronous** process that fetches entries and extracts information required
to build SCTs. When this value is exceeded, TesseraCT returns `429 - Too Many Requests`
to subsequent **duplicate** `add-*` requests only.  Non-duplicate `add-*`
requests are not impacted, and can still be processed.  This limits the number of
resources TesseraCT can spend on duplicate requests.

## Chain filtering

Chains accepted by TesseraCT can be filtered by setting the following flags:

> - `root_pem_file`: Path to the file containing root certificates that are

Reviewer: ```suggestion
- `root_pem_file`: Path to a file containing root certificates that are
```

acceptable to the log. Chains MUST be signed by one of the roots included in
this file.
- `reject_expired`: If true, TesseraCT rejects expired certificates.
- `reject_unexpired`: If true, TesseraCT rejects certificates that are either
currently valid or not yet valid.
- `ext_key_usage`: A list of comma separated [Extended Key Usages (EKU) from x509](https://pkg.go.dev/crypto/x509#ExtKeyUsage).
Certificates MUST include one or more of these EKUs to be accepted. If empty, no
filtering is applied.
- `reject_extension`: A comma separated list of X.509 extension OIDs, in dotted
string form (e.g. '2.3.4.5'). Certificates that include one or more of these extensions
will be rejected.
- `not_after_start`: Start of the range of acceptable NotAfter values,
inclusive. Leaving this unset or empty implies no lower bound to the range.
RFC3339 UTC format, e.g: 2024-01-02T15:04:05Z.
- `not_after_limit`: Cut off point of notAfter dates - only notAfter dates

> strictly *before* notAfterLimit will be accepted. Leaving this unset or empty

Reviewer: ```suggestion
strictly *before* `not_after_limit` will be accepted. Leaving this unset or empty
```

means no upper bound on the accepted range. RFC3339 UTC format, e.g:

> 2024-01-02T15:04:05Z.

Reviewer: ```suggestion
`2024-01-02T15:04:05Z`.
```

## Origin and submission prefix

The origin used in Checkpoints is specified with the `origin` flag.
The submission prefix of a log has two parts: the host and the serving path:
`https://$HOST/$PATH_PREFIX/ct/v1/...`.

> As per [static-ct-api specs](https://c2sp.org/static-ct-api), `the origin line
> MUST be the submission prefix of the log as a schema-less URL with no trailing
> slashes`. Use your frontend serving infrastructure to make sure that requests
> to these URLs are correctly routed to a TesseraCT server. TesseraCT will serve
> requests it receives regardless of their `$HOST`. However, it will match the
> $PATH_PREFIX, with the `path_prefix` flag.

Reviewer: ```suggestion
As per [static-ct-api specs](https://c2sp.org/static-ct-api):

> the origin line MUST be the submission prefix of the log as a schema-less URL with no trailing slashes.

Use your upstream serving infrastructure to make sure that requests
to these URLs are correctly routed to a TesseraCT server. TesseraCT will serve the
requests it receives regardless of their `$HOST`. However, it will expect to match requests on
`$PATH_PREFIX` as specified by the `path_prefix` flag.
```

## Memory considerations

TesseraCT's memory footprint is directly impacted by:

> - `inmemory_antispam_cache_size`: the number of entries from the antispam

Reviewer: This description isn't _quite_ right, but it's also probably good enough as is :)

database cached locally
- `batch_max_size`: the number of entries that are kept in memory before
sequenced in a batch
- [The number of cached issuers keys](https://github.com/transparency-dev/tesseract/blob/main/storage/storage.go)

> 

Reviewer: It's also heavily affected by the number of concurrent requests (because of things like context, parsed X.509 chains, goroutine stacks and overhead etc.), so `enable_awaiter` will be a "large lever" here.

> ## Multi-tenancy

Reviewer: I wonder if there's a better heading name here - mostly the paras below talk about redundancy of frontends, with only the last para saying that multi-tenancy isn't supported.

For added availability multiple TesseraCT instances can run concurently with the
same Tessera resources. Adding more instances will not necessarily increase
performance, the primary goal of concurrency is to allow for better
availability.

Multiple TesseraCT server can run concurently on AWS or GCP. It is also
possible to run concurent servers with the POSIX and Vanilla S3 + MySQL
implementations, **but** this will depend on the underlying storage systems
being used.

To run multiple logs, run multiple TesseraCT instances configured with different
Tessera resources. For simplicity, it is not possible to serve multiple logs
from a single TesseraCT instance.

## Platform specific configuration

### AWS

TesseraCT expects the databases configured with the `db_name` and
`antispam_db_name` flags to be located in the same Aurora DB cluster.

### GCP

#### VM

By default, TesseraCT exports OpenTelemetry metrics and traces to GCP
infrastructure. It is not currently possible to opt-out of this.

> When running TesseraCT locally on a VM OpenTelemetry exporters

Reviewer: ```suggestion
When running TesseraCT on a GCE VM, OpenTelemetry exporters
```
